Implement `create_sample` and `create_dynamic_sample` (#4373)
OpenDDS/OpenDDS: docs/devguide/xtypes.rst

=======================================

``DynamicData`` allows writing code that is much more generic than using the IDL-generated types.
There are still reasons to use the IDL-generated types though.
They are more convenient to use and are more likely to be used in existing code.
You can still make use of the flexibility of ``DynamicData`` though by either interchanging with or wrapping IDL-generated types.

Interchanging Data using ``create_sample`` and ``create_dynamic_sample``
------------------------------------------------------------------------

.. versionadded:: 3.29.0

The XTypes specification defines ``create_sample`` and ``create_dynamic_sample`` on the ``TypeSupport`` for every :term:`topic type` to interchange ``DynamicData`` and IDL-generated types.
In addition to this, OpenDDS adds ``create_sample_rc`` and ``create_dynamic_sample_rc`` to indicate failure and provide a more consistent interface.
The standard methods will only warn if there was an error and can't indicate that an error occured.

``create_sample`` and ``create_sample_rc``
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

``create_sample`` takes a ``DynamicData`` and returns the IDL-generated type.

> One thing to note about this method is the C++ to IDL specification requires that returns like this can be either returned on the stack or the heap depending on if they are :omgspec:`"variable-length" <cpp03:5.11 Mapping for Structured Types>`.

Reviewer: ```suggestion
One thing to note about this method is the IDL-to-C++ specification requires that returns like this can be either returned on the stack or the heap depending on if they are :omgspec:`"variable-length" <cpp03:5.11 Mapping for Structured Types>`.
```

Author: Why do I keep doing that?

For example if there's a topic type ``Foo`` that doesn't contain a ``vector`` or ``string``, then:

.. code-block:: omg-idl

  struct Foo {
    int32 bar;
  };

.. code-block:: cpp

  FooTypeSupport_var type_support = new FooTypeSupportImpl();
  DynamicData_var data = /* ... */;

  Foo foo = type_support->create_sample(data);

> If the type contains a ``vector`` or ``string`` anywhere, then then it returns a pointer that's meant to go inside a ``_var`` class:

Reviewer: ```suggestion
If the type contains a ``sequence`` or ``string`` anywhere, then then it returns a pointer that's meant to go inside a ``_var`` class:
```
Are there other IDL constructs that can be variable-length and can appear in a DDS topic type?

Author: If everything in the specs, `tao_idl`, and OpenDDS was ideal and consistent, then probably `map`. [Everything else](https://www.omg.org/spec/CPP/1.3/PDF#G6.69275) doesn't look relevant to DDS.

Reviewer: How about putting a comment in the rst (if it supports comments) that map should be included here when it's ready?

  Foo foo;
  const DDS::ReturnCode_t rc = type_support->create_sample_rc(foo, data);
  if (rc != DDS::RETCODE_OK) /* (handle error) */;

``create_dynamic_sample`` and ``create_dynamic_sample_rc``
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

``create_dynamic_sample`` takes the IDL-generated type and returns a ``DynamicData``.

.. code-block:: cpp

  FooTypeSupport_var type_support = new FooTypeSupportImpl();
  Foo foo = /* ... */;

  DynamicData_var data = type_support->create_dynamic_sample(foo);

``create_dynamic_sample_rc`` returns a ``ReturnCode_t`` to indicate failure:

.. code-block:: cpp

  FooTypeSupport_var type_support = new FooTypeSupportImpl();
  Foo foo = /* ... */;

  DynamicData_var data;
  const DDS::ReturnCode_t rc = type_support->create_dynamic_sample_rc(data, foo);
  if (rc != DDS::RETCODE_OK) /* (handle error) */;

.. _xtypes-dynamic-data-adapter:

Wrapping Data using ``DynamicDataAdapter``
------------------------------------------

.. versionadded:: 3.29.0 (but unofficially present in its current form in 3.24.0)

``DynamicDataAdapter`` is an OpenDDS-specific implementation of ``DynamicData`` that wraps an IDL-generated type.

> This allows direct interaction instead of copying all the data everytime.

Reviewer: ```suggestion
This allows direct interaction instead of copying all the data every time.
```

..
    Sect<16.8.1>

* IDL ``map`` type

* IDL ``bitmask`` type

> * Struct and union inheritance

Author: Unless I'm missing something, I don't think union inheritance is in the specs.

Reviewer: It's proposed for future versions https://issues.omg.org/issues/spec/IDL#issue-43778
That distinction was probably missed when this part of the devguide was first authored
